### DISCUSSION TITLE
Steps for next Releases

### DISCUSSION THREAD

**Discussion by svennissel** [otros-systems/otroslogviewer#626] [General]
At the moment I have no idea witch are the next step for bigger releases. Witch function will be needed? What administrative stuff must be done?

**Version 1.5.x**
* We can update Java Version to 17
* We can update used libraries
* We can remove scalar? 
* Getting better Wiki documentation
* Remove not working features?
* Code quality. Findbugs?
* More Tests?

What are the important thinks for the next time? How can you help? How can I help?

**Comment by benlazaro** [comment]
I've work on some projects where we developed components that would run on an Apache Karaf container, which is based on OSGi. For dealing with how OSGi loads/handles components I had to understand the way it manages versions, which I found it to be the way many frameworks use in their release process.
In OLV we're use the 3 number versioning format, for example: **1.2.3**, so we could follow the following convention, which I learned using Karaf:

1. **Major version**: Increase this version number when non-backwards compatible changes are included. Since this is an application and not a framework that could break backwards compatibility when updated to it, I would suggest to do this kind of release when we drop functionality from OLV, and/or when a full redesign is done. If we do decide to do this, that means we would need to schedule any changes that drop functionality from OLV to be scheduled in a "major milestone". We could also bundle some 10-15 pull requests which add new functionality and release it as a major version, for example. But in any case, releasing a major version requires more planning.
2. **Minor version**: This number should be increased when new features are added. This can also include bug fixes, but as long a we have a new feature (even if it's a single one), we should release a new minor version. This means that whatever is set to be release with a minor version should be added to a "minor milestone"
3. **Micro version (AKA "patch version")**: This type of release version includes bug fixes/patches, minor changes that do not change the behavior of the application, but just improve the overall performance. I'm not sure if we want to create a milestone for this. We could just cut a release every certain amount of time and include any bug fixes we have available at that time. Say we release once every month and we add all the bug fixes that passed peer review and made it to the master branch.

So if we decide to adopt this convention for our release process, I think the list from the original post would look like this:

- **We can update Java Version to 17** (This one kind of breaks backwards compatibility with older version of the JDK, but I've seen these type of updates released in a minor version, so we can add it to milestone 1.5.0, otherwise 2.0.0)
- **We can update used libraries** (milestone 1.5.0)
- **We can remove scalar?** (This would be replacing Scala code with Java code, so as long as the functionality remains the same in the formatters code we can add it to milestone 1.5.0, but it would be best to schedule it for milestone 2.0.0)
- **Getting better Wiki documentation** (I don't know how the wiki is released, but if we decide to reformat it or add a lot of new documentation pages, I would add this to milestone 1.5.0, or we can do small improvements and release them in micro versions instead)
- **Remove not working features?** (This would probably be better if we add it to a major release, even if the features are not working)
- **Code quality. Findbugs?** (As long as we don't change any functionality and this is only code improvements, I think milestone 1.5.0 should be fine)
- **More Tests?** (Unit tests are usually going to require changes in the code, and some are going to go as far as redesigning some components, so as long as the functionality remains the same, this could go into minor releases, like different milestones)

Having a convention like this helps the users to know what type of changes to expect from each release version.
That's my two cents on versioning.

**Comment by svennissel** [comment]
**We can update Java Version to 17**
I think we use the Version 2.0.0 for major changes from the perspective of the end user. Something like new UI design or big new features. 
So I want to put Java 17 build to release 1.5. Than Java 6 is 1.3, Java 8 1.4 and Java 11 is 1.5.

**We can update used libraries**
I created a issue

**We can remove scalar**
I do not know at the moment what functionality is implemented in scalar. But a the last time I want to build the project with eclipse the scalar plugin for eclipse did not work. I have create a task to check this. It is very important for me that every developer can easy start to support OtrosLogViewer if he want

**Getting better Wiki documentation**
The Wiki is not released. The main problem is I don't know how a user with no permissions can make something like a pull request for the wiki.

**Remove not working features**
That is right. And the best solution is to fix the features. Did you have a overview of not working features?

**Code quality. Findbugs**
I have create a task. I don't know at the moment a easy way for local execution with a nice report/integration. But I think there will be a way.
